File: ios/PlaytomicFoundation/storage/FileStorage.swift
//
//  FileStorage.swift
//  Anemone SDK
//
//  Created by Angel Luis Garcia on 08/02/2017.
//  Copyright © 2017 Syltek Solutions S.L. All rights reserved.
//

import Foundation

public class FileStorage: IFileStorage {
    public enum Exception: Error {
        case notFound, wrongUrl
    }

    let path: String

    public init(directory: FileManager.SearchPathDirectory) {
        let paths = NSSearchPathForDirectoriesInDomains(directory, .userDomainMask, true)
        self.path = paths[0]
    }

    public func writeData(_ data: Data, file: String) throws {
        let url = try fileURL(name: file)
        try data.write(to: url, options: Data.WritingOptions.atomic)
    }

    public func writeDataAsync(_ data: Data, file: String) -> Promise<Void> {
        Promise(executeInBackground: true) { fulfill, _ in
            try self.writeData(data, file: file)
            fulfill(())
        }
    }

    public func readData(file: String) -> Data? {
        guard let url = try? fileURL(name: file) else {
            return nil
        }
        return try? Data(contentsOf: url)
    }

    public func readDataAsync(file: String) -> Promise<Data> {
        Promise(executeInBackground: true) { fulfill, reject in
            if let data = self.readData(file: file) {
                fulfill(data)
            } else {
                reject(Exception.notFound)
            }
        }
    }

    public func removeData(file: String) throws {
        let url = try fileURL(name: file)
        try FileManager.default.removeItem(at: url)
    }

    public func removeDataAsync(file: String) -> Promise<Void> {
        Promise(executeInBackground: true) { fulfill, _ in
            try self.removeData(file: file)
            fulfill(())
        }
    }

    private func fileURL(name: String) throws -> URL {
        guard let url = URL(string: "file://\(path)/\(name)") else {
            throw Exception.wrongUrl
        }
        return url
    }
}

File: android/app/src/main/java/com/playtomic/test/foundation/storage/FileStorage.kt
package com.playtomic.foundation.storage

import com.playtomic.foundation.extension.tryOrNull
import com.playtomic.foundation.promise.Promise
import java.io.*
import java.lang.Exception

/**
 * Created by agarcia on 09/02/2017.
 */

class FileStorage(private val directory: String) : IFileStorage {
    sealed class Exception : java.lang.Exception {

        constructor()

        constructor(message: String?) : super(message)

        constructor(cause: Exception) : super(cause)

        object notFound : Exception()
        object wrongUrl : Exception()
    }

    @Throws(IOException::class)
    override fun writeData(data: ByteArray, file: String) {
        try {
            synchronized(this) {
                val stream = FileOutputStream(getFile(file))
                stream.write(data)
                stream.flush()
                stream.close()
            }
        } catch (e: FileNotFoundException) {
            throw IOException(e)
        }
    }

    override fun readData(file: String): ByteArray? {
        try {
            synchronized(this) {
                val stream = tryOrNull { FileInputStream(getFile(file)) }
                val data = tryOrNull { stream?.readBytes() }
                stream?.close()
                return data
            }
        } catch (e: Exception) {
            return null
        }
    }

    private fun getFile(name: String): File = File(directory, name)

    override fun writeDataAsync(data: ByteArray, file: String): Promise<Unit> =
        Promise(executeInBackground = true) { fulfill, _ ->
            writeData(data, file)
            fulfill(Unit)
        }

    override fun readDataAsync(file: String): Promise<ByteArray> =
        Promise(executeInBackground = true) { fulfill, reject ->
            val data = readData(file)
            if (data != null) {
                fulfill(data)
            } else {
                reject(IOException())
            }
        }

    override fun removeData(file: String) {
        getFile(name = file).delete()
    }

    override fun removeDataAsync(file: String): Promise<Unit> =
        Promise(executeInBackground = true) { fulfill, _ ->
            this.removeData(file = file)
            fulfill(Unit)
        }
}
